keps/sig-storage/3476-volume-group/README.md

```diff
@@ -0,0 +1,1647 @@
+
+#### ModifyVolume
+
+ModifyVolume CSI RPC was considered earlier to add/remove one volume to/from a group at a time but it was removed because ModifyVolumeGroup CSI RPC was added.
+
+A new MODIFY_VOLUME capability will be added to support this.
+It indicates that the controller plugin supports modifying a volume.
+
+```
+  rpc ModifyVolume(ModifyVolumeRequest)
+    returns (ModifyVolumeResponse) {
+        option (alpha_method) = true;
+    }
+```
+
+This RPC is called when an existing volume is added to an existing volume group or when a volume is removed from the volume group.
+A volume group id parameter will be in the ModifyVolumeRequest for an add request.
+A volume group id parameter will not be in the ModifyVolumeRequest for a delete request.
+If user requests to add an existing volume to a consistency group, but the CSI driver cannot fulfill the request because the existing volume is placed on a different storage pool from the consistency group, then the CSI driver MUST return failure.
+This RPC MUST be idempotent.
+
+```
+message ModifyVolumeRequest {
+      string volume_id = 1;
+
+      // This field is OPTIONAL.
+      repeated string volume_group_id = 2 [(alpha_field) = true];
+
+      // Secrets required by plugin to complete modify volume request.
+      // This field is OPTIONAL. Refer to the `Secrets Requirements`
+      // section on how to use this field.
+      map<string, string> secrets = 3 [(csi_secret) = true];
+}
+```
+External-provisioner will be modified so that modifying PVC by adding VolumeGroupName will trigger a ModifyVolume call (a new CSI controller RPC) to CSI driver.
+
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Other
+    - Describe the mechanism:
+      The external volume group and group snapshot controllers do not have a
+      feature gate because they are out of tree.
+      It is enabled when these external controller sidecars are deployed with the CSI driver.
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+      From the controller side, it only affects the external controller sidecars.
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+      No.
+
+* **Does enabling the feature change any default behavior?**
+  Yes. Enabling the feature can allow a new PVC to be created and added to a VolumeGroup. Enabling the feature can also allow a VolumeSnapshot to be created as part of the VolumeSnapshotGroup.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+  Yes. All VolumeGroup and VolumeGroupSnapshot API objects need to be deleted before this feature can be truly disabled.
```

Comment:
The sidecar also needs to be removed, correct? Can there be unusual situations where the CRDs are deleted but the sidecar is still present, for VolumeGroups that were created when the feature was enabled?

Reply:
Yes. First all the new API objects need to be deleted.  Then the new controllers for VolumeGroup and VolumeGroupSnapshot can be removed, and external-provisioner and external-snapshotter controllers/sidecars need to be downgraded to a version without this feature.
Let me clarify that.